# Pandas (2)

Материалы:
* Макрушин С.В. "Лекция 2: Библиотека Pandas"
* https://pandas.pydata.org/docs/user_guide/index.html#
* https://pandas.pydata.org/docs/reference/index.html
* Уэс Маккини. Python и анализ данных

## Задачи для совместного разбора

1. Загрузите данные из файла `sp500hst.txt` и обозначьте столбцы в соответствии с содержимым: `"date", "ticker", "open", "high", "low", "close", "volume"`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('sp500hst.txt', sep=',', header=None, names=["date", "ticker", "open", "high", "low", "close", "volume"])

In [3]:
df.head()

date ticker   open     high     low  close  volume
0  20090821      A  25.60  25.6100  25.220  25.55   34758
1  20090824      A  25.64  25.7400  25.330  25.50   22247
2  20090825      A  25.50  25.7000  25.225  25.34   30891
3  20090826      A  25.32  25.6425  25.145  25.48   33334
4  20090827      A  25.50  25.5700  25.230  25.54   70176

2. Посчитайте количество уникальных цифр, которые используются каждой строке в столбце volume.

3. Для каждой строки рассчитайте разность между значениями high и low, если индекс столбца нечетный, и разность между close и high в противном случае.

4. Рассчитайте суммарный объем торгов для для одинаковых значений тикеров.

5. Загрузите данные из файла sp500hst.txt и обозначьте столбцы в соответствии с содержимым: "date", "ticker", "open", "high", "low", "close", "volume". Добавьте столбец с расшифровкой названия тикера, используя данные из файла `sp_data2.csv` . В случае нехватки данных об именах тикеров корректно обработать их.

## Лабораторная работа №2.2

__Данная работа является продолжением ЛР №2. Для начала работы загрузите таблицы (см. задание 1.1)__

In [4]:
recipes = pd.read_csv('recipes_sample.csv')

In [5]:
reviews = pd.read_csv('reviews_sample.csv')

In [6]:
reviews.rename(columns={'Unnamed: 0': 'review_id'}, inplace=True)

In [7]:
reviews = reviews.set_index('review_id').sort_values(by='review_id')

In [8]:
reviews.head()

user_id  recipe_id        date  rating  \
review_id                                              
5               52282     120345  2005-05-21       4   
6              124416     120345  2011-08-06       0   
7          2000192946     120345  2015-05-10       2   
8               76535     134728  2005-09-02       4   
9              273745     134728  2005-12-22       5   

                                                      review  
review_id                                                     
5          very very sweet. after i waited the 2 days i b...  
6          Just an observation, so I will not rate.  I fo...  
7          This recipe was OVERLY too sweet.  I would sta...  
8                                                 Very good!  
9                                     Better than the real!!

### Применение функций к pd.Series и pd.DataFrame

4.1 Напишите функцию, которая переводит минуты в формат "XhYm". Примените эту функцию к столбцу `minutes` таблицы `recipes` (без перезаписи исходного столбца) при помощи метода `map`.

In [9]:
def m_to_hm(m: int) -> str:
    return(f'{m//60}h{m%60}m')

In [10]:
m_to_hm_all = list(map(m_to_hm,list(recipes.minutes)))

In [11]:
m_to_hm_all[:10]

['1h30m',
 '0h10m',
 '0h30m',
 '0h45m',
 '0h25m',
 '0h15m',
 '0h28m',
 '0h5m',
 '0h45m',
 '0h40m']

4.2 На основе таблицы `recipes` создайте таблицу, которая содержит только текстовые столбцы (используйте метод `select_dtypes`).  Примените к каждому элементу этой таблицы строковый метод `str.capitalize` при помощи метода `applymap`, не удаляя пропуски.

In [12]:
import copy

In [13]:
recipes_copy = copy.copy(recipes.select_dtypes('object'))

In [14]:
recipes_copy.iloc[0][0].capitalize()

'George s at the cove  black bean soup'

In [15]:
recipes_copy.dtypes

name           object
submitted      object
description    object
dtype: object

In [16]:
pd.__version__

'1.0.5'

`na_action` не доступен в текущей версии pandas

```recipes_copy.applymap(lambda x: str(x).capitalize(),na_action='ignore')```

In [17]:
recipes_copy.isna().sum()

name             0
submitted        0
description    623
dtype: int64

In [18]:
recipes_copy.fillna('NaN', inplace=True)

In [19]:
recipes_copy.applymap(lambda x: str(x).capitalize())

name   submitted  \
0             George s at the cove  black bean soup  2002-10-25   
1                Healthy for them  yogurt popsicles  2003-07-26   
2                      I can t believe it s spinach  2002-08-29   
3                              Italian  gut busters  2002-07-27   
4          Love is in the air  beef fondue   sauces  2004-02-23   
...                                             ...         ...   
29995  Zurie s holey rustic olive and cheddar bread  2007-11-25   
29996          Zwetschgenkuchen  bavarian plum cake  2009-08-24   
29997   Zwiebelkuchen   southwest german onion cake  2004-11-03   
29998                                   Zydeco soup  2012-08-29   
29999        Cookies by design   cookies on a stick  2008-04-15   

                                             description  
0      An original recipe created by chef scott meska...  
1      My children and their friends ask for my homem...  
2                These were so go, it surprised even me.  
3      My sister-in-law made these for us at a family...  
4      I think a fondue is a very romantic casual din...  
...                                                  ...  
29995  This is based on a french recipe but i changed...  
29996  This is a traditional fresh plum cake, thought...  
29997  This is a traditional late summer early fall s...  
29998  This is a delicious soup that i originally fou...  
29999  I've heard of the 'cookies by design' company,...  

[30000 rows x 3 columns]

4.3 Напишите функцию, которая принимает на вход серию `pd.Series` и для серий, содержащих текстовые данные, возвращает максимальную длину строк в ней, а для числовых серий возвращает минимальный элемент в этой серии. Примените данную функцию к каждому столбцу таблицы `recipes` при помощи метода `apply`.

In [20]:
def get_stats(x: pd.Series) -> int:
    if(x.dtype == 'O'):
        return max(x.str.len()) 
    if((x.dtype == 'int64') or (x.dtype == 'float64')):
        return min(x)
    

assert get_stats(pd.Series(['a', 'bbbb', 'ccc'])) == 4
assert get_stats(pd.Series([1, 3, 2])) == 1

Не получилось

In [21]:
recipes.apply(get_stats)

name                83.0
id                   NaN
minutes              NaN
contributor_id       NaN
submitted           10.0
n_steps              NaN
description       6321.0
n_ingredients        NaN
dtype: float64

### Группировки таблиц `pd.DataFrame`

5.1 Посчитайте количество рецептов, представленных каждым из участников (`contributor_id`). Какой участник добавил максимальное кол-во рецептов?

In [22]:
count_of_rec = recipes.groupby(by='contributor_id').count()
count_of_rec

name   id  minutes  submitted  n_steps  description  \
contributor_id                                                        
1530               5    5        5          5        2            2   
1533             186  186      186        186      114           70   
1534              50   50       50         50       27           15   
1535              40   40       40         40       21           38   
1538               8    8        8          8        5            3   
...              ...  ...      ...        ...      ...          ...   
2001968497         2    2        2          2        0            2   
2002059754         1    1        1          1        1            1   
2002234079         1    1        1          1        1            1   
2002234259         1    1        1          1        1            1   
2002247884         1    1        1          1        1            1   

                n_ingredients  
contributor_id                 
1530                        3  
1533                      133  
1534                       36  
1535                       29  
1538                        6  
...                       ...  
2001968497                  1  
2002059754                  1  
2002234079                  1  
2002234259                  0  
2002247884                  1  

[8404 rows x 7 columns]

Количество рецептов, представленных каждым из участников = **8408**

In [23]:
count_of_rec.loc[count_of_rec['name'].idxmax()]

name             421
id               421
minutes          421
submitted        421
n_steps          254
description      421
n_ingredients    291
Name: 89831, dtype: int64

Участник с индексом **89831**

**5.2** Посчитайте средний рейтинг к каждому из рецептов. Для скольких рецептов отсутствуют отзывы? Обратите внимание, что отзыв с нулевым рейтингом или не заполненным текстовым описанием не считается отсутствующим.

Средний рейтинг по каждому из рецептов

In [24]:
reviews.groupby('recipe_id').mean()

user_id    rating
recipe_id                        
48         5.054750e+04  1.000000
55         5.328002e+05  4.750000
66         1.116224e+08  4.944444
91         2.353162e+05  4.750000
94         7.743438e+05  5.000000
...                 ...       ...
536547     2.002045e+09  5.000000
536610     9.269990e+05  0.000000
536728     6.906230e+05  4.000000
536729     6.908992e+05  4.750000
536747     1.501687e+09  0.000000

[28100 rows x 2 columns]

In [25]:
len(recipes.groupby('id').count().index) - len(reviews.groupby('recipe_id').count().index)

1900

Для **1900** рецептов отсутствуют отзывы

**5.3** Посчитайте количество рецептов с разбивкой по годам создания.

In [26]:
recipes["submitted"] = pd.to_datetime(recipes["submitted"])

In [27]:
recipes.groupby(recipes["submitted"].dt.year).count()

name    id  minutes  contributor_id  submitted  n_steps  \
submitted                                                            
1999        275   275      275             275        275      164   
2000        104   104      104             104        104       50   
2001        589   589      589             589        589      360   
2002       2644  2644     2644            2644       2644     1652   
2003       2334  2334     2334            2334       2334     1481   
2004       2153  2153     2153            2153       2153     1309   
2005       3130  3130     3130            3130       3130     1995   
2006       3473  3473     3473            3473       3473     2210   
2007       4429  4429     4429            4429       4429     2752   
2008       4029  4029     4029            4029       4029     2546   
2009       2963  2963     2963            2963       2963     1821   
2010       1538  1538     1538            1538       1538      961   
2011        922   922      922             922        922      565   
2012        659   659      659             659        659      422   
2013        490   490      490             490        490      344   
2014        139   139      139             139        139       93   
2015         42    42       42              42         42       30   
2016         24    24       24              24         24       16   
2017         39    39       39              39         39       24   
2018         24    24       24              24         24       15   

           description  n_ingredients  
submitted                              
1999               150            203  
2000                61             69  
2001               507            417  
2002              2406           1873  
2003              2228           1646  
2004              2142           1526  
2005              3124           2196  
2006              3469           2437  
2007              4427           3090  
2008              4027           2841  
2009              2959           2081  
2010              1538           1086  
2011               922            668  
2012               659            447  
2013               490            365  
2014               139             89  
2015                42             27  
2016                24             16  
2017                39             28  
2018                24             15

**5.4** Напишите функцию, которая принимает на вход таблицу (аналогичную `recipes` по набору столбцов), и возвращает `True` в том случае, если в столбце `minutes` присутствуют только значения, меньшие либо равные 10. Сгруппируйте таблицу `recipes` по полю `contributor_id` и для каждого участника выясните, справедливо ли, что все его рецепты занимают не более 10 минут.

Не делал

In [28]:
def has_only_fast_recipes(x: pd.DataFrame) -> bool:
    pass

assert not has_only_fast_recipes(
    pd.DataFrame(
        {
            "name": {0: "george s", 1: "healthy"},
            "id": {0: 44123, 1: 67664},
            "minutes": {0: 90, 1: 10},
            "contributor_id": {0: 35193, 1: 91970},
            "submitted": {0: "2002-10-25", 1: "2003-07-26"},
            "n_steps": {0: np.nan, 1: np.nan},
            "description": {0: "123", 1: "zxc"},
            "n_ingredients": {0: 18.0, 1: np.nan},
        }
    )
)
assert has_only_fast_recipes(
    pd.DataFrame(
        {
            "name": {0: "george s", 1: "healthy"},
            "id": {0: 44123, 1: 67664},
            "minutes": {0: 7, 1: 5},
            "contributor_id": {0: 35193, 1: 91970},
            "submitted": {0: "2002-10-25", 1: "2003-07-26"},
            "n_steps": {0: np.nan, 1: np.nan},
            "description": {0: "123", 1: "zxc"},
            "n_ingredients": {0: 18.0, 1: np.nan},
        }
    )
)

NameError: name 'np' is not defined

### Объединение таблиц `pd.DataFrame`

6.1 При помощи объединения таблиц, создайте `DataFrame`, состоящий из четырех столбцов: `id`, `name`, `user_id`, `rating`. Рецепты, на которые не оставлен ни один отзыв, должны отсутствовать в полученной таблице. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и попытавшись найти строку, соответствующую этому рецепту, в полученном `DataFrame`.

In [29]:
recipes['id']

0         44123
1         67664
2         38798
3         35173
4         84797
          ...  
29995    267661
29996    386977
29997    103312
29998    486161
29999    298512
Name: id, Length: 30000, dtype: int64

In [30]:
dataframe_6_1 = pd.merge(recipes[["id","name"]], 
                         reviews[["recipe_id", "user_id", "rating"]], 
                         how='inner', 
                         left_on="id", 
                         right_on='recipe_id').drop("recipe_id",axis=1)

In [31]:
dataframe_6_1

id                                    name  user_id  rating
0        44123   george s at the cove  black bean soup    34206       5
1        44123   george s at the cove  black bean soup    76503       5
2        44123   george s at the cove  black bean soup   743566       5
3        67664      healthy for them  yogurt popsicles    32772       5
4        67664      healthy for them  yogurt popsicles   161653       5
...        ...                                     ...      ...     ...
126691  486161                             zydeco soup   133174       5
126692  486161                             zydeco soup   305531       5
126693  486161                             zydeco soup   724631       5
126694  486161                             zydeco soup   226863       5
126695  298512  cookies by design   cookies on a stick   804234       1

[126696 rows x 4 columns]

6.2 При помощи объединения таблиц и группировок, создайте `DataFrame`, состоящий из трех столбцов: `recipe_id`, `name`, `review_count`, где столбец `review_count` содержит кол-во отзывов, оставленных на рецепт `recipe_id`. У рецептов, на которые не оставлен ни один отзыв, в столбце `review_count` должен быть указан 0. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и найдя строку, соответствующую этому рецепту, в полученном `DataFrame`.

Не делал

6.3. Выясните, рецепты, добавленные в каком году, имеют наименьший средний рейтинг?

Не делал

### Сохранение таблиц `pd.DataFrame`

7.1 Отсортируйте результат выполнения задания 6.1 в порядке убывания величины столбца `id` и сохраните результаты в csv файл. 

Не делал

7.2 Воспользовавшись `pd.ExcelWriter`, cохраните результаты 6.1 и 6.2 в файл: на лист с названием `Рецепты с оценками` сохраните результаты выполнения 6.1; на лист с названием `Количество отзывов по рецептам` сохраните результаты выполнения 6.2.

Не делал